From b3b496302d5c17aef588f3a15f9f612295825390 Mon Sep 17 00:00:00 2001
From: "sameer.mulla" <sameer.mulla@lge.com>
Date: Sat, 22 Aug 2020 15:28:41 +0530
Subject: [PATCH] caching of avdtp delay reporting

:Release Notes:
Cache Delay reporting capability of remote SEID

:Detailed Notes:
Delay reporting procedure is not happening due to caching
of remote SEID sometimes. Current caching of SEID is not storing
delay reporting capability. Added caching implementaion
of delay reporting along with SEID caching information

:Testing Performed:
Builded and Tested

:QA Notes:
NA

:Issues Addressed:
[PLAT-117479] Cache Delay reporting capability of remote SEID

Upstream Status: Pending
---
 profiles/audio/a2dp.c  | 13 ++++++++-----
 profiles/audio/avdtp.c | 10 +++++++++-
 profiles/audio/avdtp.h |  5 ++++-
 3 files changed, 21 insertions(+), 7 deletions(-)

diff --git a/profiles/audio/a2dp.c b/profiles/audio/a2dp.c
index a5590b24c..84f568054 100644
--- a/profiles/audio/a2dp.c
+++ b/profiles/audio/a2dp.c
@@ -1931,6 +1931,7 @@ static void load_remote_sep(struct a2dp_channel *chan, GKeyFile *key_file,
 	for (; *seids; seids++) {
 		uint8_t type;
 		uint8_t codec;
+		uint8_t delay_reporting;
 		GSList *l = NULL;
 		char caps[256];
 		uint8_t data[128];
@@ -1944,8 +1945,8 @@ static void load_remote_sep(struct a2dp_channel *chan, GKeyFile *key_file,
 		if (!value)
 			continue;
 
-		if (sscanf(value, "%02hhx:%02hhx:%s", &type, &codec,
-								caps) != 3) {
+		if (sscanf(value, "%02hhx:%02hhx:%02hhx:%s", &type, &codec, &delay_reporting,
+								caps) != 4) {
 			warn("Unable to load Endpoint: seid %u", rseid);
 			g_free(value);
 			continue;
@@ -1968,7 +1969,7 @@ static void load_remote_sep(struct a2dp_channel *chan, GKeyFile *key_file,
 
 		caps_add_codec(&l, codec, data, size / 2);
 
-		rsep = avdtp_register_remote_sep(chan->session, rseid, type, l);
+		rsep = avdtp_register_remote_sep(chan->session, rseid, type, l, !!delay_reporting);
 		if (!rsep) {
 			warn("Unable to register Endpoint: seid %u", rseid);
 			continue;
@@ -2594,6 +2595,8 @@ static struct queue *a2dp_select_eps(struct avdtp *session, uint8_t type,
 static void store_remote_sep(void *data, void *user_data)
 {
 	struct a2dp_remote_sep *sep = data;
+	uint8_t delay_reporting = avdtp_get_delay_report(sep->sep);
+	DBG("store_remote_sep avdt_get_delay_report(sep->sep) %d", delay_reporting);
 	GKeyFile *key_file = (void *) user_data;
 	char seid[4], value[256];
 	struct avdtp_service_capability *service = avdtp_get_codec(sep->sep);
@@ -2603,8 +2606,8 @@ static void store_remote_sep(void *data, void *user_data)
 
 	sprintf(seid, "%02hhx", avdtp_get_seid(sep->sep));
 
-	offset = sprintf(value, "%02hhx:%02hhx:", avdtp_get_type(sep->sep),
-						codec->media_codec_type);
+	offset = sprintf(value, "%02hhx:%02hhx:%02hhx:", avdtp_get_type(sep->sep),
+						codec->media_codec_type, delay_reporting);
 
 	for (i = 0; i < service->length - sizeof(*codec); i++)
 		offset += sprintf(value + offset, "%02hhx", codec->data[i]);
diff --git a/profiles/audio/avdtp.c b/profiles/audio/avdtp.c
index 0e075f9ff..a99a4ba77 100644
--- a/profiles/audio/avdtp.c
+++ b/profiles/audio/avdtp.c
@@ -3199,6 +3199,11 @@ uint8_t avdtp_get_type(struct avdtp_remote_sep *sep)
 	return sep->type;
 }
 
+gboolean avdtp_get_delay_report(struct avdtp_remote_sep *sep)
+{
+	return sep->delay_reporting;
+}
+
 struct avdtp_service_capability *avdtp_get_codec(struct avdtp_remote_sep *sep)
 {
 	return sep->codec;
@@ -3223,7 +3228,9 @@ struct avdtp_service_capability *avdtp_service_cap_new(uint8_t category,
 struct avdtp_remote_sep *avdtp_register_remote_sep(struct avdtp *session,
 							uint8_t seid,
 							uint8_t type,
-							GSList *caps)
+							GSList *caps,
+							gboolean delay_reporting
+)
 {
 	struct avdtp_remote_sep *sep;
 	GSList *l;
@@ -3238,6 +3245,7 @@ struct avdtp_remote_sep *avdtp_register_remote_sep(struct avdtp *session,
 	sep->type = type;
 	sep->media_type = AVDTP_MEDIA_TYPE_AUDIO;
 	sep->caps = caps;
+	sep->delay_reporting = delay_reporting;
 
 	for (l = caps; l; l = g_slist_next(l)) {
 		struct avdtp_service_capability *cap = l->data;
diff --git a/profiles/audio/avdtp.h b/profiles/audio/avdtp.h
index ad2cb9bcb..c48ee2432 100644
--- a/profiles/audio/avdtp.h
+++ b/profiles/audio/avdtp.h
@@ -226,12 +226,15 @@ struct avdtp_service_capability *avdtp_service_cap_new(uint8_t category,
 struct avdtp_remote_sep *avdtp_register_remote_sep(struct avdtp *session,
 							uint8_t seid,
 							uint8_t type,
-							GSList *caps);
+							GSList *caps,
+							gboolean delay_reporting);
 
 uint8_t avdtp_get_seid(struct avdtp_remote_sep *sep);
 
 uint8_t avdtp_get_type(struct avdtp_remote_sep *sep);
 
+gboolean avdtp_get_delay_report(struct avdtp_remote_sep *sep); 
+
 struct avdtp_service_capability *avdtp_get_codec(struct avdtp_remote_sep *sep);
 
 int avdtp_discover(struct avdtp *session, avdtp_discover_cb_t cb,
-- 
2.14.1

